zipextimporter.py

```diff
@@ -42,138 +43,227 @@
+                    return spec
+                return None
+            if mi.is_ext:
+                spec = ModuleSpec(fullname, self.zipextimporter, origin=mi.path)
+                if mi.is_package:
+                    spec.submodule_search_locations = [mi.path.rpartition("\\")[0]]
             else:
-                super().exec_module(module)
+                try:
+                    loader = self.zipimporter
+                except AttributeError:
+                    loader = self
+                spec = spec_from_loader(fullname, loader, is_package=mi.is_package)
+            return spec
+
+    if hasattr(zipimporter, "load_module"):
+        def load_module(self, fullname):
+            # will never enter here, raise error for developers
+            raise NotImplementedError("load_module() is not implemented, "
+                                      "use create_module() instead.")
+
+    def create_module(self, spec):
+        spec._set_fileattr = True  # has_location, use for reload
+        fullname = spec.name
+        origin = spec.origin
+
+        # PEP 489 multi-phase initialization / Export Hook Name
+        name = fullname.rpartition(".")[2]
+        try:
+            name.encode("ascii")
+        except UnicodeEncodeError:
+            initname = "PyInitU_" + name.encode("punycode") \
+                                        .decode("ascii").replace("-", "_")
+        else:
+            initname = "PyInit_" + name
+
+        mod = _memimporter.import_module(fullname, origin, initname,
+                                         self.get_data, spec)
+        if self.verbose:
+            print(f"import {fullname} # loaded from zipfile {self.archive}",
+                  file=sys.stderr)
+        return mod
+
+    def exec_module(self, module):
+        # all has been done in create_module(), also skip importlib.reload()
+        pass
+
+    ## ====================== improves compatibility ======================
+    def get_code(self, fullname):
+        mi = self._get_module_info(fullname, _raise=True)
+        if not mi.is_ext:
+            return self.zipimporter.get_code(fullname)
+
+    def get_source(self, fullname):
+        mi = self._get_module_info(fullname, _raise=True)
+        if not mi.is_ext:
+            return self.zipimporter.get_source(fullname)
+
+    def get_filename(self, fullname):
+        mi = self._get_module_info(fullname, _raise=True)
+        if not mi.is_ext:
+            return self.zipimporter.get_filename(fullname)
+        return mi.path
+
+    def is_package(self, fullname):
+        mi = self._get_module_info(fullname, _raise=True)
+        return mi.is_package
+    ## ====================================================================
 
     def __repr__(self):
-        return "<%s object %r>" % (self.__class__.__name__, self.archive)
+        return super().__repr__().replace("zipimporter", "ZipExtensionImporter")
+
+
+def install(hook=hasattr(zipimporter, "_files")):
```

Comment:
I am a little confused... the class `zipimporter` has never an attribute `_files`. So, what is the difference between this and `hook=False`? 

In addition, could you comment a little on the two installation methods? Why was the old one (`_install_hook`) not good enough? Why monkey patching `zipimport` should be better? I find inserting in the `path_hooks` is much more elegant and stable than messing with code that we cannot truly control...

Reply:
There is no difference, `hasattr(...)` expression was kept as memento. If unadvisable, please remove it.

I think monkey patch `zipimport.zipimporter` is better way, will not break codes which base on  `zipimport.zipimporter`. Because `zipextimporter` is not a official lib module, there is no compatibility at all.

Reply:
I do not understand this. In the current implementation, code that is based on `zipimporter` will work out-of-the-box provided that the author changes the way they initialize the importer. There should be no need for monkey patching the original code.